### DISCUSSION TITLE
How to custom extractors error response?

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#354] [Q&A] [answered]
I want to customize the content of the response, how can I get the information provided by `JsonRejection`.

```rust
impl IntoResponse for ServerError {
    type Body = Full<Bytes>;
    type BodyError = Infallible;

    fn into_response(self) -> Response<Self::Body> {
        let (status_code, message) = match self {
            Self::ValidationError(_) => {
                let message = format!("Input validation error: [{}]", self).replace("\n", ", ");
                tracing::debug!("{}", message);
                (StatusCode::BAD_REQUEST, message)
            }
            Self::AxumJsonError(err) => {
                let message = format!("{:?}", err);
                (err.into_response().status(), message)
            }
            Self::ReInitializedError => (StatusCode::FORBIDDEN, self.to_string()),
            _ => {
                tracing::error!("Internal Server Error: [{}]", self.to_string());
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "Internal Server Error".to_string(),
                )
            }
        };

        let payload = Json(json!({ "error": message }));
        (status_code, payload).into_response()
    }
}
```

**Comment by davidpdrsn** [comment, marked as answer]
I don't quite follow what you mean. [`JsonRejection`](https://docs.rs/axum/0.2.5/axum/extract/rejection/enum.JsonRejection.html) is an enum so you can `match` on each variant and handle the case. Similarly to whats shown [here](https://docs.rs/axum/0.2.5/axum/index.html#optional-extractors).

**Reply by ghost** [reply]
Sorry, I didn't describe it clearly.
```rust
Self::AxumJsonError(err) => {
    let message = match err {
        JsonRejection::InvalidJsonBody(e) => {
            // Item `0` is private, I want to get it
            format!("{}", e.0)
        }
        _ => todo!()
    };
    (err.into_response().status(), message)
}
```

**Reply by Svenskunganka** [reply]
The `e` variable is an [`InvalidJsonBody`](https://docs.rs/axum/0.2.5/axum/extract/rejection/struct.InvalidJsonBody.html) struct that implements both `Debug` and `Display`. Does using `e` directly rather than `e.0` for `format!` do what you want?  
Although the `Display` impl of the `InvalidJsonBody` struct will only write `Failed to parse the request body as JSON` to the formatter, which I guess isn't what you want?

Instead, you can get the `e.0` field by calling [`e.source()`](https://docs.rs/axum/0.2.5/axum/extract/rejection/struct.InvalidJsonBody.html#method.source) as implemented [in this macro](https://github.com/tokio-rs/axum/blob/f8a0d81d79453c5364105361948bd439d9dd3c90/src/macros.rs#L108-L112).

**Reply by ghost** [reply]
> The `e` variable is an [`InvalidJsonBody`](https://docs.rs/axum/0.2.5/axum/extract/rejection/struct.InvalidJsonBody.html) struct that implements both `Debug` and `Display`. Does using `e` directly rather than `e.0` for `format!` do what you want? Although the `Display` impl of the `InvalidJsonBody` struct will only write `Failed to parse the request body as JSON` to the formatter, which I guess isn't what you want?
> 
> Instead, you can get the `e.0` field by calling [`e.source()`](https://docs.rs/axum/0.2.5/axum/extract/rejection/struct.InvalidJsonBody.html#method.source) as implemented [in this macro](https://github.com/tokio-rs/axum/blob/f8a0d81d79453c5364105361948bd439d9dd3c90/src/macros.rs#L108-L112).

thank you very much!!!
